### PR TITLE
Feat/key did secp256r1 now includes curves secp384r1 and secp521r1 as well

### PR DESCRIPTION
**Pull request by bshambaugh**
Feat/key did secp256r1 now includes curves secp384r1 and secp521r1 as well

## Description

Code has been extended to include Secp384r1 and Secp521r1 and refactored to put common functions into a separate file nist_weierstrass_common .

I will double check this code in the next day or so. Right now, a nap.

### CODE REVIEW

File: packages/key-did-resolver/README.md [hunk 1]
```diff
@@ -6,6 +6,325 @@
...
+
+The did:key id is the multicodec name + the compressed public key expressed as a Uint8Array (byte array) then encoded with base58btc and expressed as a string.
+The syntax of a did:key is did:key:id.
+
+Compressed keys are the X coordinate of the public key with a prefix that depends on the sign of the Y curve coordinate. The prefix is '02' if even and '03' if odd.
+
+Example code for creating a did:key for P-256, P-384, and P-521 is here:
+https://github.com/bshambaugh/did-key-creator 
+
+## Code
```
Comment (oed) [inline packages/key-did-resolver/README.md:28]:
```suggestion
#### Code
```

File: packages/key-did-resolver/README.md [hunk 2]
```diff
@@ -6,6 +6,325 @@
...
+```
+// Usage from cloned GitHub Repository:
+// import * as keyDIDResolver from '../js-ceramic/packages/key-did-resolver/lib/index.js';
+
+import keyDIDResolver from 'key-did-resolver'
+
+const resolverRegistry = keyDIDResolver.getResolver();
+
+const resolve = resolverRegistry.key;
+
```
Comment (oed) [inline packages/key-did-resolver/README.md:38]: Can you make the example more like this?
https://www.npmjs.com/package/@ceramicnetwork/3id-did-resolver

i.e. show it's usage together with the `did-resolver` pkg

File: packages/key-did-resolver/src/__tests__/secp384r1.test.ts [hunk 3]
```diff
@@ -0,0 +1,292 @@
+// Brent Shambaugh <brent.shambaugh@gmail.com>. 2021.
+
+import varint from "varint"
+import multibase from "multibase"
```
Comment (oed) [inline packages/key-did-resolver/src/__tests__/secp384r1.test.ts:4]: the `multibase` package has been deprecated, could you update to use `multiformats`?
See https://github.com/ceramicnetwork/js-ceramic/pull/1862 for inspiration.

**Review by oed — CHANGES_REQUESTED** [review]
Thanks for your work on this! We are currently in the process on migrating to ES2020 and ESM, once that's done I will give this a much more thorough review and we can merge this. 

Are we currently using the test vectors from https://github.com/w3c-ccg/did-method-key/tree/main/test-vectors ?

File: packages/key-did-resolver/src/__tests__/secp384r1.test.ts (hunk 3 above)
Comment (oed) [reply packages/key-did-resolver/src/__tests__/secp384r1.test.ts:4]: Also, can you target the `release/2.x-progress` branch?

**Comment by bshambaugh** [comment]
> Thanks for your work on this! We are currently in the process on migrating to ES2020 and ESM, once that's done I will give this a much more thorough review and we can merge this.
> 
> Are we currently using the test vectors from https://github.com/w3c-ccg/did-method-key/tree/main/test-vectors ?

When matching against https://github.com/w3c-ccg/did-method-key/tree/main/test-vectors I have. 
P-256:
https://gist.github.com/bshambaugh/f175c39df239951121e25f2152601de5

P-384:
https://gist.github.com/bshambaugh/65b52d0db051ff91864847a67e505fc8

P-521:
https://gist.github.com/bshambaugh/4980c0c6f1df473448ce55f1d358963e

I believe these may be already in the tests.

This was a good exercise because it did turn up something. Although the x,y coordinates in base64url resolve like the test vectors, I notice my json-ld response is in a slightly different form. I first notice that I have no @base property.

I was scared the other day. I went to https://did.key.transmute.industries/generate/secp521r1 and noticed a difference when I threw in the did:keys and tried to resolve to the base64url points, but as far as the test vectors concerned the keys and the resolved points have been consistent. Also, what I put in from the library elliptic to create a did:key is what I get out. I think this is something I have to talk to Orie about. The w3c ccg test vectors have been accepted by the community. I have not delved into the specifics of his code. I was just playing around on his site after I found it looking for did:keys.

**Comment by bshambaugh** [comment]
Adding a @base part of the json-ld would require a change to all key:did curves. secp256k1, ed25519, P-256, P-384, P-521.
I'll focus on your other comments and revisit this later.

**Comment by bshambaugh** [comment]
This is somewhat related. Is https://github.com/ceramicnetwork/key-did-provider-ed25519 like the key-did-provider library that you want to have? https://github.com/ceramicnetwork/key-did-provider-ed25519 . I noticed that this is the only one that is available on npm. https://github.com/ceramicnetwork/key-did-provider-secp256k1 exists on github, but not npm. I think https://github.com/bshambaugh/did-key-creator is okay, but it does not have JWE or JWK, and it does not expose a constructor, just a few functions that allow one to create did:keys.

File: packages/key-did-resolver/README.md (hunk 2 above)
Comment (bshambaugh) [reply packages/key-did-resolver/README.md:38]: import {DID} from 'dids'
import KeyResolver from 'key-did-resolver'
import {Ed25519Provider} from 'key-did-provider-ed25519'
import {randomBytes} from '@stablelib/random'

// this stuff might be a bit beyond the scope
import { CeramicClient } from '@ceramicnetwork/http-client'
const API_URL = "https://ceramic-clay.3boxlabs.com"
const ceramic = new CeramicClient(API_URL)
//

const seed = randomBytes(32) //  32 bytes with high entropy
const provider = new Ed25519Provider(seed)

const did = new DID({ provider, resolver: KeyResolver.getResolver() })
const auth = await did.authenticate()
console.log('auth is')
console.log(auth) // this spits out the did:key

// create JWS
const { jws, linkedBlock } = await did.createDagJWS({ hello: 'world' })

// create JWE
const jwe = await did.createDagJWE({ very: 'secret' }, [did.id])

console.log('jwe is:');
console.log(jwe);

// decrypt JWE
const decrypted = await did.decryptDagJWE(jwe)

console.log('decrypted is:');
console.log(decrypted);

// this stuff may be a bit beyold the scope
ceramic.did = did
ceramic.did.setProvider(provider)
//

Comment (bshambaugh) [reply packages/key-did-resolver/README.md:38]: That's my latest draft.

**Comment by bshambaugh** [comment]
This pull request replaces pull request #1441 .

File: packages/key-did-resolver/README.md [hunk 4]
```diff
@@ -6,6 +6,253 @@
...
+### Usage
+This code includes support for the curves Ed25519, Secp256k1, Secp256r1 (P-256), Secp384r1 (P-384), and Secp521r1 (P-521) which follow the test vectors at:
+[https://github.com/w3c-ccg/did-method-key/tree/main/test-vectors](https://github.com/w3c-ccg/did-method-key/tree/main/test-vectors)
+
+This code has been tested with the following `did:key`[^1] providers:
+|  Curve              | Repositry                                                      |
+| ------------------- | -------------------------------------------------------------- |
+| Ed25519             | https://github.com/ceramicnetwork/key-did-provider-ed25519     |
+| Secp256k1           | https://github.com/ceramicnetwork/key-did-provider-secp256k1   |
+| P-256, P-384, P-521 | https://github.com/bshambaugh/did-key-creator                  |
```
Comment (bshambaugh) [inline packages/key-did-resolver/README.md:18]: Random thought. This may need to live over in https://github.com/ceramicstudio .

**Comment by bshambaugh** [comment]
> > Thanks for your work on this! We are currently in the process on migrating to ES2020 and ESM, once that's done I will give this a much more thorough review and we can merge this.
> > Are we currently using the test vectors from https://github.com/w3c-ccg/did-method-key/tree/main/test-vectors ?
> 
> When matching against https://github.com/w3c-ccg/did-method-key/tree/main/test-vectors I have. P-256: https://gist.github.com/bshambaugh/f175c39df239951121e25f2152601de5
> 
> P-384: https://gist.github.com/bshambaugh/65b52d0db051ff91864847a67e505fc8
> 
> P-521: https://gist.github.com/bshambaugh/4980c0c6f1df473448ce55f1d358963e
> 
> I believe these may be already in the tests.
> 
> This was a good exercise because it did turn up something. Although the x,y coordinates in base64url resolve like the test vectors, I notice my json-ld response is in a slightly different form. I first notice that I have no @base property.
> 
> I was scared the other day. I went to https://did.key.transmute.industries/generate/secp521r1 and noticed a difference when I threw in the did:keys generated into my code and tried to resolve to the base64url points, but as far as the test vectors concerned the keys and the resolved points have been consistent. Also, what I put in from the library elliptic to create a did:key is what I get out. I think this is something I have to talk to Orie about. The w3c ccg test vectors have been accepted by the community. I have not delved into the specifics of his code. I was just playing around on his site after I found it looking for did:keys.

Okay, Ories code always seems to match up with my x,y coordinates when I generate starting from https://did.key.transmute.industries/ and then click on the SECP521R1 link but whenever I click regenerate at https://did.key.transmute.industries/generate/secp521r1 I do not get alignment.

File: packages/key-did-resolver/README.md (hunk 2 above)
Comment (bshambaugh) [reply packages/key-did-resolver/README.md]: See the README for the latest. I think I was headed down the wrong path originally, but it made me think about did-jwt, did-jwt-vc, and future directions I may want to take https://github.com/bshambaugh/did-key-creator .

**Comment by OR13** [comment]
Its possible there are bugs, code reviews and independent implementations help trace them down.... One of the best ways is importing and exporting them, and cross testing with libraries you trust... historically I have tried to use https://www.npmjs.com/package/jose for that.

How can I help on this PR?

**Comment by bshambaugh** [comment]
> Its possible there are bugs, code reviews and independent implementations help trace them down. :)
> 
> One of the best ways is importing and exporting them, and cross testing with libraries you trust... historically I have tried to use https://www.npmjs.com/package/jose for that.
> 
> How can I help on this PR?

Okay. I'll comb through stuff a bit more. I know the key length of P-521 varies, so I allowed for a hex length between 132 and 134. I saw 132 in your code. I'm a bit of a rookie here. I don't mean to burden you. I alerted you, so i guess that is something.

**Comment by OR13** [comment]
@bshambaugh you are not a rookie, I see your stuff ; )

P-521 generally suffers from a lack of redundant implementations... there is a good argument to be made that nobody should use it... but imo it best not to conflate "standard support" with "security best practices"... keep both and separate the layers cleanly... You should also beware of this:

https://groups.google.com/g/mozilla.dev.security.policy/c/7O34-DmZeC8?pli=1

If you can prove that our library is doing something wrong, with a minimal preproduction, we are happy to address (or you are welcome to PR it in yourself).

**Comment by bshambaugh** [comment]
Here is my 'back of the envelope' calculation to create did:keys from the npm elliptic library and then resolve:
[](https://gist.github.com/bshambaugh/833604bd3f59a75b6cf5ce4d56515316)

This code produces did:keys that resolve to what the committed code produces. The code shows that the BigINT x,y values are the same before and after compression as well as before did:key creation (straight from the elliptic library output).

I know of no way to prove that my P-521 commit is incorrect. TODO: revisit Transmute's did:key.

**Comment by OR13** [comment]
> I know of no way to prove that my P-521 commit is incorrect. 

The way to do that is to show that you handle the test vectors here: https://w3c-ccg.github.io/did-method-key/#p-521

If the test vectors are wrong, you can open a PR to change them...

The spec is the source of truth, not individual implementations... we should fix the spec first, if there is a problem... then open issues on any implementations that are not spec conformant.

**Comment by bshambaugh** [comment]
> > I know of no way to prove that my P-521 commit is incorrect.
> 
> The way to do that is to show that you handle the test vectors here: https://w3c-ccg.github.io/did-method-key/#p-521
> 
> If the test vectors are wrong, you can open a PR to change them...
> 
> The spec is the source of truth, not individual implementations... we should fix the spec first, if there is a problem... then open issues on any implementations that are not spec conformant.

I am matching the x,y base64url values for each did:key. I have some questions about the spec, so I will ask there {https://github.com/w3c-ccg/did-method-key/}.

**Comment by oed** [comment]
@bshambaugh We just merged the Ceramic 2.0 branch into develop (with ESM support 🎉 ). Can you please merge develop into this branch and solve any conflicts? After that I will take anther look at this PR and hopefully merge it!

**Comment by bshambaugh** [comment]
I'm trying to track this TypeError error message inconsistency down:
scr-1643062004

**Comment by bshambaugh** [comment]
> I'm trying to track this TypeError error message inconsistency down: scr-1643062004

My I index.test.ts file was much larger too.

**Comment by bshambaugh** [comment]
I'm investigating the media type the key-did-resolver outputs: 
scr-1643137913

**Comment by bshambaugh** [comment]
> I'm investigating the media type for the key-did-resolver tests (e.g: https://github.com/bshambaugh/js-ceramic/blob/feat/key-did-secp256r1/packages/key-did-resolver/src/__tests__/__snapshots__/index.test.ts.snap#L374-L410)

To the best of my knowledge, what I have seems okay. I talked to Orie Steele. 
https://lists.w3.org/Archives/Public/public-credentials/2022Jan/0213.html

**Comment by OR13** [comment]
Ya'll should probably try to send someone to the VC WG calls. We discussed this issue today:

https://github.com/w3c/vc-wg-charter/issues/46

**Comment by bshambaugh** [comment]
"Final" thought:
I recall that in March or May of last year things were coming along, and I think I had some semblance of P-256 support. Then, due to discussions on 

https://github.com/w3c-ccg/did-method-key/issues/29

I added some support for P-384 and P-521 to help out. There also was not consensus on whether to use compressed keys yet, so support for uncompressed and raw keys (without 04 prefix) for lack of better word was added in addition to compressed.

Discussions here drove https://github.com/multiformats/multicodec/blob/master/table.csv to be updated to be compressed. As everything was in flux, this arose:

https://github.com/w3c-ccg/did-method-key/issues/37

I kept the legacy support for uncompressed and compressed keys, but made a note in the readme:
"Compressed2 forms of P-256, P-384, and P-521 are preferred. 3" with reference to a footnote.

I know initially you did not like the idea of uncompressed and raw keys, but I explained the context and you were okay with it even though I think we are both in agreement that they could cause confusion.

The short reason is:
Lauren got back to me and wondered why things were dragging on with https://github.com/bshambaugh/BlinkyProject  and when I was going to launch . Truth be told I was having a few sluggish months, so to catch back up I'm not worrying about perfection. 

On the CCG list there appears to be some discussion about independent review of NIST curves used for this application (maybe Ver Creds?) by SRI (if I recall). There also is discussion about P-384 for support of large organizations using hardware security modules.  I know Ceramic's focus has not been on the NIST curves. The whole PR has been driven by Blinky. 

I'm now looking at did-jwt to finish stuff up (and calling this okay for now). I see: https://github.com/decentralized-identity/did-jwt/issues/170 with mention to reorganize the code. 

https://github.com/decentralized-identity/did-jwt/pull/212 is where this is going on. All of the updates are coming in from my fork. I'm driving through the did-jwt code trying to make the best sense I can. Basically, I'm setting this up to add a ES256 signer so I can work with an essentially remote HSM (https://github.com/decentralized-identity/did-jwt/blob/master/docs/guides/index.md#creating-custom-signers-for-integrating-with-hsm). 

Due to splitting stuff up into different files, I imagine also that ES384/p384, ES512/p521 could be easily added (the hope) as well as RSA and BBS+? by interested parties. I wasn't there when stuff was spun out of uPort (I presume)..so I'm just running through guessing trying to get something operational.

**Comment by bshambaugh** [comment]
I'll forget. The VCWG is moving toward secp384r1 support for legacy HSM. https://lists.w3.org/Archives/Public/public-credentials/2022Jan/0186.html . Also, I recall that large cloud providers like Azure support all three NiST curves. I propose helping out with the WG if desired.

File: packages/key-did-resolver/src/secp256r1.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
+// Brent Shambaugh <brent.shambaugh@gmail.com>. 2021.
+
+import * as u8a from 'uint8arrays'
+import * as bigintModArith from 'bigint-mod-arith'
+
+import * as nist_weierstrass_common from './nist_weierstrass_common'
```
Comment (ukstv) [inline packages/key-did-resolver/src/secp256r1.ts:6]:
```suggestion
import * as nist_weierstrass_common from './nist_weierstrass_common.js'
```

File: packages/key-did-resolver/src/secp384r1.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
+// Brent Shambaugh <brent.shambaugh@gmail.com>. 2021.
+
+import * as u8a from 'uint8arrays'
+import * as bigintModArith from 'bigint-mod-arith'
+
+import * as nist_weierstrass_common from './nist_weierstrass_common'
```
Comment (ukstv) [inline packages/key-did-resolver/src/secp384r1.ts:6]:
```suggestion
import * as nist_weierstrass_common from './nist_weierstrass_common.js'
```

**Comment by bshambaugh** [comment]
Item #3 at https://github.com/ceramicnetwork/js-ceramic/pull/1441#issuecomment-1031515117 . Is still a work in progress.

I made a note: "Would someone be willing to check this out for me? At present I believe I am implicitly importing as an ESM [1]  by using the "type":"module" in my package.json . However, my reviewer still reports problems on their machine. https://github.com/ceramicnetwork/js-ceramic/pull/1441#issuecomment-1031515117 .  An alternative would be to use .mjs extensions instead of  "type":"module" in my package.json  .The PR https://github.com/ceramicnetwork/js-ceramic/pull/1884 is the new pull request, and my commits are coming from my fork at https://github.com/bshambaugh/js-ceramic/tree/feat/key-did-secp256r1 . I'm not sure how to sandbox this. It works fine on my machine, but I could change to  .mjs just 'cause.[1] https://dev.to/bnb/implicit-esm-in-node-js-with-type-module-np "

Useful Refs:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import

https://github.com/juanelas/bigint-mod-arith
"or import (JavaScript ES module):

import * as bigintModArith from 'bigint-mod-arith'

The appropriate version for browser or node is automatically exported.
You can also download the [IIFE bundle](https://raw.githubusercontent.com/juanelas/bigint-mod-arith/master/dist/bundles/iife.js), the [ESM bundle](https://raw.githubusercontent.com/juanelas/bigint-mod-arith/master/dist/bundles/esm.min.js) or the [UMD bundle](https://raw.githubusercontent.com/juanelas/bigint-mod-arith/master/dist/bundles/umd.js) and manually add it to your project, or, if you have already imported bigint-mod-arith to your project, just get the bundles from node_modules/bigint-mod-arith/dist/bundles/."

**Comment by ukstv** [comment]
Hey, @bshambaugh . Made an effort to rebase your branch onto the latest develop, and played with it. Apparently, all the tests are ok (even runs in ESM mode), despite what CircleCI says.

I am convinced, this could be merged.

**Comment by bshambaugh** [comment]
Could the  Circle CI issue be for another package? see the raw circle ci output for this PR: https://circleci.com/api/v1.1/project/github/ceramicnetwork/js-ceramic/5462/output/111/0?file=true&allocation-id=6203ea99d400852d4e16b211-0-build%2F5EAE6C79

**Pull request merged**